### ISSUE TITLE
Proposal: Reconsider how we authenticate package location changes

### ISSUE THREAD

**Issue by thomashoneyman** [purescript/registry-dev#331]
The PureScript manifest includes a field called `repository`, which indicates either a GitHub owner + repo or a URL to an arbitrary Git repository online, along with an optional `subdir` key that notes where in the repository the manifest can be found.

https://github.com/purescript/registry/blob/62b93f1cf16d97520979925b5a0fcc525f6c77fc/v1/Manifest.dhall#L22-L23

However, the only time we ever use the repository key (and thus the `Repo` type) is via the Registry API:

https://github.com/purescript/registry/blob/62b93f1cf16d97520979925b5a0fcc525f6c77fc/v1/Operation.dhall#L7-L13

We _have_ to take the URL and `subdir` key before we can fetch the package in an `Addition`; without this information, we can't retrieve the manifest. We don't have to take this information in for an `Update` because we can look up the package's location via the registry index...unless the user changed the package's repository, in which case they have to submit an `Addition` instead of an `Update`, despite the package already being registered.

I think it would be simpler to store the package location in the package metadata, not the manifest. We need the package location in order to retrieve its manifest, so storing it in the manifest provides no benefit. Worse, the manifest location can easily be false; a user could send one location via the API, while storing a different location in the manifest. We could validate that the manifest matches with the API call, but this is one more brittle check in an already-complicated pipeline. (We've forgotten to include this check all along, for example!).

Storing the location in the version metadata, by comparison, means the metadata tells you exactly where the registry built _that_ version from. There is only a single source of truth: the API call, which is then put into the metadata, verified by whether we could actually retrieve the package and its manifest or not. And it's one less piece of information that has to be maintained in a manifest file by the user, or their package manager, or by us.

Nothing is lost with this change, as far as I can tell. Package managers don't use the repository information because they download packages from the registry. If they _did_ want the information, it's still available via the registry metadata, which package managers already access to get other information about packages. Package manager configurations are unaffected; they still need to tell the registry the package location, same as before.

Therefore, I'm proposing:

1. We remove the `repository` key from the manifest, and we add it to the version metadata
2. We continue to take the repository via the registry API, exactly as we do now.

**Comment by thomashoneyman** [comment]
I forgot we already _do_ store this information in the version metadata! All the more reason to remove it from the manifest, where it is just duplicated and possibly incorrect information. My proposal then is actually:

1. We remove the `repository` key from the manifest
2. We otherwise make no changes

**Comment by f-f** [comment]
I feel that this proposal, #333 and #332 all come from a desire to reduce complexity, but there are reasons why we ended up with this design and why I think these proposals would end up increasing complexity, let me try to illustrate why.

Let's consider the problem of "how do we allow package owners to change the URL where the Registry fetches the package from".

One solution for that problem is what you propose in #332 and #333 that is "always provide the URL, and validate the git commit if that URL changes".
The biggest drawback of this solution is that it depends on `git` signatures. Note that we currently _don't_ depend on `git` at all!
I would like to try to avoid a dependency unless we absolutely have to. This is because depending on it would prevent us from eventually adding things like a `TarballProvider` to the [possible places where we fetch packages from](https://github.com/purescript/registry/blob/master/v1/Repo.dhall) - right now they are only git-thingies, but nothing is preventing us from adding more providers that are not git-like (and in fact the `GitHub` provider _is_ actually just fetching tarballs)

Another solution to that problem is what we specced together [about six months ago](https://github.com/purescript/registry/issues/155#issuecomment-900905245) (and did not implement yet, which is probably why you found out that some of this stuff is unused and may look unnecessary), which looks something like this:
- when registering a package the package manager takes the URL from the `repository` key of the Manifest and puts that into an `Addition`
- we store that URL in the Metadata, and that's where we will get the URL to fetch all the subsequent versions published, which will all come through an `Update` (note that `Additions` get transformed in `Updates` if a package is already registered)
- whenever a new version goes through the pipeline, we look at the URL coming from the Manifest. _If that's different_ from the one in the Metadata, _then_ we update the one in the Metadata with the new one, so that _the next_ fetches will go through that URL

In short, the reason why we need to store the repo/url information in the Manifest is so that owners can change the URL where the registry goes to fetch the future versions.

This is safe for the same reasoning that makes the SSH-signatures-dance safe, which is that we can only trust data coming from new releases, and we can't trust data coming in input from the API.

I much prefer this solution to the one that uses git signatures because I don't want to incur on a hard dependency on git.

**Comment by thomashoneyman** [comment]
> I feel that this proposal, #333 and #332 all come from a desire to reduce complexity, but there are reasons why we ended up with this design and why I think these proposals would end up increasing complexity

These proposals do come from a desire to reduce complexity, and thank you for providing some extra context around the design. Part of why I find them complex is because my mental model of how we handle trust, and the tools we use for verification, aren't totally in line with yours. It's good we can have a discussion and work this out!

I'd like to try and thoroughly explain my understanding of the channels we trust, how we verify information, and what tools we incur hard dependencies on. From there I hope we can narrow down where we are diverging from each other.

I've written a book below, but I promise it's not a waste of time to read! All this information needs to end up in the spec one way or another, and this can serve as a first draft.

### Trust

One common thread we're talking through is trust. There are several components to trust in the registry:

* What actions in the registry are sensitive?
* What channels of communication do we have with package owners? What channels do we trust?
* How do we verify an owner if they need to take a sensitive action, but we can't communicate with them via a trusted channel?

**Sensitive Actions**

There are two sensitive actions in the registry (actions that require proof of ownership): changing a package repo and unpublishing a package version. No other action requires proof of ownership. For example, we don't require any proof of ownership to add a new package to the registry.

**Channels of Communication**

We have a few ways to communicate with package owners:

1. Users can reach out to the registry team directly at registry@purescript.org, or open issues on the registry repository on GitHub
2. Users can send payloads via the registry API
3. Users can make changes to their manifest file in their repository

Not all of these channels are trusted; we must verify package ownership when communicating in an untrusted channel.

GitHub issues and the registry email address are untrusted channels. They're primarily for DMCA takedown notices, security notices, and other special circumstances, not for the two sensitive actions we support.

The registry API is also considered an untrusted channel. That means that sensitive actions cannot be taken directly via the API:

> Note that we have to take the URL from the metadata because we can't trust whatever is coming from the API. This is why we can't process URL updates from the API [^1].

Right now we generally do a good job of not trusting the API. But we _do_ trust some information provided via the API; for example, we take in the package name via the API and use it without verifying it against the manifest [^2]. (Presumably we should remove this field and just take the value from the manifest?)

The only trusted channel is the manifest file located in a repository:

> Our only trusted channel of communication with package authors/owners/maintainers (all of which I'll call "owners" from now) is a release. I.e. whenever we go to a repo and package a specific ref then we establish a chain of trust with whoever pushed that ref.

This means we do not need to authenticate sensitive actions if we can process the request via the manifest. For example, if a user wants to process a URL update, we can do it in a trusted way by having them update the URL in their manifest, publish a new release, and then ask to publish that release. We'll pick up the location change in a trusted way.

**Verification**

Finally, there's the case when there we need to communicate with a package owner over an untrusted channel, and so we need them to verify they are indeed the package owner(s). For that, we've settled on having them provide their public key via their repository (a trusted channel), and to use their private key to sign messages to us over the untrusted channel [^3].

### Tooling Dependencies

Another common thread to our discussion is what tooling we expect packages to use. For example, in our discussion on the package maintainers issue you proposed that we can use SSH keys to sign payloads when we need to prove package ownership [^3]. This introduces our first hard dependency: `ssh`.

I have also assumed a hard dependency on `git`. I haven't seen this as a problem; as you stated about SSH, the overwhelming majority of people already use it. We've also heavily oriented the registry's design around git repositories. Not all packages are on GitHub -- some are on Sourcehut [^4] -- but _all_ PureScript packages I'm aware of are Git repositories. That said:

> Note that we currently don't depend on git at all!
> 
> I much prefer this solution to the one that uses git signatures because I don't want to incur on a hard dependency on git. [^5]

It's technically true that we don't require packages to use `git`, but the registry is certainly oriented around Git. For example, we only accept packages from GitHub right now. The only other "provider" we have support for is an arbitrary Git URL. And our communication during the design process has implicitly assumed packages use Git:

> One of the main points of this proposal is that it's completely agnostic, there are absolutely no GitHub-specific bits in here and any git forge will do. We literally just need to be able to fetch the repo in some way. [^6]

> Our only trusted channel of communication with package authors/owners/maintainers (all of which I'll call "owners" from now) is a release. I.e. whenever we go to a repo and package a specific ref then we establish a chain of trust with whoever pushed that ref. We can do this because GitHub already established an authentication from the real person to the online place where we fetch stuff from. [^7]

If we should be VCS-agnostic, then we almost certainly need to take another pass through our decisions with an eye on other systems: SVN, Darcs, and Pijul come to mind. Possibly we've remained agnostic enough that we can add support for these other systems when the time comes, but this is completely untested and I at least haven't been designing with these other systems in mind.

---

All that is my understanding, and it leads to a few questions that I still need to work out. They include:

### The API

The API is an untrusted channel of communication with the user. This means that we must verify sensitive actions sent via the API (ie. `Unpublish`) and it means we should always prefer information via the manifest to information via the API.

**Issue 1: Accepting too much information via the API**

When a package is registered we accept the package name via the API and use it to register the package [^2], even though this information is also present in the manifest (and may be different in the manifest!).

It seems that if the API is not at trustworthy channel then we should prefer the manifest contents. For example we could take only the repository location in an `Addition`, use that to fetch the manifest, and then read the manifest's `name` field.

**Issue 2: URL changes are unclear**

Prior to your response in #331, I thought we were using the `Addition` operation both to register a package and to change its location later on. You helped make it clear that this operation is only ever for initial registration, and it's also obvious to me now that we couldn't use this operation for URL changes because it disrupts the continuity of the repo that we use for trust.

I've closed #331 because it's clear we shouldn't merge `Update` and `Addition`.

However, the process for changing a repo location via an `Update` is still a little unclear to me. Here are the steps from your comment over in #331:

> * when registering a package the package manager takes the URL from the `repository` key of the Manifest and puts that into an `Addition`
> * we store that URL in the Metadata, and that's where we will get the URL to fetch all the subsequent versions published, which will all come through an `Update` (note that `Additions` get transformed in `Updates` if a package is already registered)

These two steps seem fine to me -- other than that the package manager will probably take the URL from its own configuration format, generate the manifest, and then put the same value into an `Addition`.

> * whenever a new version goes through the pipeline, we look at the URL coming from the Manifest. _If that's different_ from the one in the Metadata, _then_ we update the one in the Metadata with the new one, so that _the next_ fetches will go through that URL

This part is a little unclear to me. This implies that you have to take these steps to move your repository:

1. You commit and tag a new version in which you change the `repository` field in your manifest.
2. You transfer the repository to its new location (for example, you move your repo from your account to `purescript-contrib`).
3. You ask the registry to perform an `Update`, giving it the old repository. The registry will see that the old repository has a different repo specified in the manifest, so it will go fetch _that_ repo and publish it at the ref you asked for.

If I've summarized this correctly, then this shuffle sidesteps the authentication question, but it has problems. It feels like poor UX to have to tag two separate releases and maintain the repository in two locations until the registry can process the change. Here's why:

First, I'm sure folks will inadvertently transfer their repositories to their new location without publishing a new version of their repo at the old location that updates the repository field in the manifest. I'm sure people will transfer the repo, then try to publish from the new location, which will fail.

Second, it feels like a poor user experience if users commit at the old location and ask the registry to publish before they've transferred the repository to its new location. The registry would see the updated URL and go to retrieve the new location, only for it to not exist yet.

Third -- and this follows from the second point -- it's a bit of a poor user experience to require two separate releases in the first place, and to require the repository to be in both locations at once until the registry finishes publishing. It's not really a problem with GitHub because it'll do automatic redirects (transfer your repo, and the old link will still work). But we certainly shouldn't design this with GitHub's redirects in mind -- we don't even want a hard dependency on Git, let alone GitHub!

It seems like we should probably treat repository transfers the same as an `Unpublish` operation: require a signed payload proving package ownership to do it. That way you can just transfer the repo, and when you're ready to publish from the new location you do it the same way you register a package for the first time -- except this time you have to prove your package ownership.

---

I know I've written a literal book here, so I'd like to wrap up with some action items we can use to move forward.

First, we need to make a decision as to whether we are VCS-agnostic or not. If we are, then we take a pass through the registry to ensure we are not requiring packages to use Git specifically anywhere. If we are not, then we should be able to assume a hard `git` dependency and we should say that up front.

Second, we should formally state that we do not trust input via the API unless it is signed, and we defer to the manifest wherever possible. Right away we can change the `Addition` operation to only take a location and ref (we'll get the package name from the manifest). The `Update` operation can remain unchanged.

Third, we should nail down the URL change issue. My proposal is that URL change operations should be authenticated exactly the same as `Unpublish` operations (ie. by signing the payload via private SSH keys). However, please let me know if I've misrepresented the process and it's fine as-is.

[^1]: https://github.com/purescript/registry/issues/332#issuecomment-1030827472
[^2]: [The implementation of `runOperation`](https://github.com/purescript/registry/blob/1255b3a03525b1757f638e523d656ace00c1d590/ci/src/Registry/API.purs#L105-L123) and [`addOrUpdate`](https://github.com/purescript/registry/blob/1255b3a03525b1757f638e523d656ace00c1d590/ci/src/Registry/API.purs#L186-L189).
[^3]: https://github.com/purescript/registry/issues/155#issuecomment-1023767402
[^4]: https://github.com/purescript/registry/pull/204
[^5]: https://github.com/purescript/registry/issues/331#issuecomment-1030825172
[^6]: https://github.com/purescript/registry/issues/155#issuecomment-1026596616
[^7]: https://github.com/purescript/registry/issues/155#issuecomment-1024889973

**Comment by f-f** [comment]
> we take in the package name via the API and use it without verifying it against the manifest. (Presumably we should remove this field and just take the value from the manifest?)

In the `Addition` we need it to register the new package (I don't think we enforce that it's the same as in the `Manifest` in the non-legacy codepath), and in `Update` we need it to lookup which package we're talking about.
The only "untrusted" usage is the first, but that's fine because we're just adding a new package to the list of the published ones. The API caller doesn't gain any access if they don't control the location in the first place.

> For example, we only accept packages from GitHub right now. 

As I noted above our implementation is basically "fetch a tarball from GitHub", no `git` required

> And our communication during the design process has implicitly assumed packages use Git

Sure, because as you noted all the packages we're importing are on GitHub, and saying "fetch from the next release" is more immediate than "fetch from any place on the internet, with any protocol, as long as we have some kind of schema that allows us to figure out where to fetch the next thing from", if you're trying to explain something else.

> I at least haven't been designing with these other systems in mind

I have been designing while trying to keep agnostic to the protocol we use to fetch stuff, even though in many places we do say "git" rather than "whatever protocol" (names are hard), so we could do a pass removing this terminology if you'd like

**Comment by f-f** [comment]
> It seems that if the API is not at trustworthy channel then we should prefer the manifest contents. For example we could take only the repository location in an Addition, use that to fetch the manifest, and then read the manifest's name field.

While this might work for new packages, I doubt it would work for legacy imports?
IIRC the names in the manifests are often garbage.

> 1. You commit and tag a new version in which you change the `repository` field in your manifest.
> 2. You transfer the repository to its new location (for example, you move your repo from your account to `purescript-contrib`).
> 3. You ask the registry to perform an Update, giving it the old repository. The registry will see that the old repository has a different repo specified in the manifest, so it will go fetch that repo and publish it at the ref you asked for.

After step (1) the Registry already has picked up the change in the URL when you publish this version (I assume you meant "publish" rather than "tag" - we don't need git tags at all! We pick the version from the `Manifest`), so after step (2) the new owners are ready to publish from the new location.
There's no step (3) (also: we don't accept a `URL` in the `Update` operation)

> If I've summarized this correctly, then this shuffle sidesteps the authentication question, but it has problems. It feels like poor UX to have to tag two separate releases and maintain the repository in two locations until the registry can process the change. Here's why:

Since there's no step 3, we are looking at only one place and only one version published for doing the change. This is about the shortest protocol I can think of in any case? (short of already having keys available and going through another channel)

> Second, it feels like a poor user experience if users commit at the old location and ask the registry to publish before they've transferred the repository to its new location. The registry would see the updated URL and go to retrieve the new location, only for it to not exist yet.
> It seems like we should probably treat repository transfers the same as an `Unpublish` operation: require a signed payload proving package ownership to do it. That way you can just transfer the repo, and when you're ready to publish from the new location you do it the same way you register a package for the first time -- except this time you have to prove your package ownership.

This sounds a bit strange to me - you mention the poor UX of the solution, and then propose something that has a strictly worse UX: using the signed payloads requires users to mess with keys (I don't know if we can pick one automatically or if we have to require that they add it manually), and likely requires to publish a new version to add such keys (since about 100% of the repos don't have those right now).
This is an additional burden that we shouldn't require of them, if we can avoid it.
However, we could _also_ allow them to use an authenticated operation to do this, just in case they move repo location before publishing a new release with the new URL.

**Comment by f-f** [comment]
Some more notes:
- in case we want to add an out-of-band operation for changing the URL through ssh signatures, we'd need to find a way to keep that in sync with the URL in the repo. E.g. if you have A in the Manifest, and change it to B through a `Signed` operation, then the next release should have B as the URL.
  We could do this with a flag in the `Metadata`, but I'd rather not have two ways of doing the same thing.
- I can't quite explain why, but it feels wrong to _not_ store in the data itself the location where the data was fetched from

**Comment by thomashoneyman** [comment]
> > For example, we only accept packages from GitHub right now.
> 
> As I noted above our implementation is basically "fetch a tarball from GitHub", no `git` required
> 
> > And our communication during the design process has implicitly assumed packages use Git
> 
> Sure, because as you noted all the packages we're importing are on GitHub, and saying "fetch from the next release" is more immediate than "fetch from any place on the internet, with any protocol, as long as we have some kind of schema that allows us to figure out where to fetch the next thing from", if you're trying to explain something else.
> 
> > I at least haven't been designing with these other systems in mind
> 
> I have been designing while trying to keep agnostic to the protocol we use to fetch stuff, even though in many places we do say "git" rather than "whatever protocol" (names are hard), so we could do a pass removing this terminology if you'd like

I'm trying to clarify what exactly we _do_ require of packages. Would you say that these are accurate requirements we impose on your package?

1. Your package must be accessible publicly on the Internet, ie. over `http` or `https`
2. Your package must use a version control system with the concept of a commit which uniquely identifies the code at a moment in time, and this information must be accessible to the registry when we fetch your package
3. We currently support the Git version control system, and have the ability to support others in the future

In addition, in the case a package owner wants to take a sensitive operation, we require them to have a public/private SSH keypair for authentication.

We don't allow "any protocol" because we at least require you to provide a ref + version when registering and updating packages. But we are pretty flexible about it! It helps me to know specifically what we plan to support -- please let me know if these three points about cover it!

**Edit:** We also have requirements on package contents, of course, like how they have to have a manifest (except for legacy packages). But that's not relevant to this part of the discussion.

**Comment by thomashoneyman** [comment]
> > we take in the package name via the API and use it without verifying it against the manifest. (Presumably we should remove this field and just take the value from the manifest?)
> 
> In the `Addition` we need it to register the new package (I don't think we enforce that it's the same as in the `Manifest` in the non-legacy codepath), and in `Update` we need it to lookup which package we're talking about. The only "untrusted" usage is the first, but that's fine because we're just adding a new package to the list of the published ones. The API caller doesn't gain any access if they don't control the location in the first place.

> > It seems that if the API is not at trustworthy channel then we should prefer the manifest contents. For example we could take only the repository location in an Addition, use that to fetch the manifest, and then read the manifest's name field.
> 
> While this might work for new packages, I doubt it would work for legacy imports? IIRC the names in the manifests are often garbage.

Ugh, what I'm describing would be problematic for legacy imports, you're totally right. It sounds like we _should_ preserve the `packageName` field to support legacy packages specifically. Still, it feels to me that we should verify that the package name provided via the API matches the one in the manifest in the non-legacy code path.

For example, what if I tell the registry to publish someone else's package, `purescript-abc`, but I send via the API the package name `purescript-xyz`? Their package will be published under the name `xyz`, even though they stated in their manifest that the package name is `abc`. The package owner doesn't necessarily publish the package, after all.

What do you think of following a rule where if we receive information via the API that is also present in the manifest, we must verify that the two fields match? For example, if we receive a repo location, go fetch the manifest there, and read in the `repository` field, should we verify that the repository field is actually the same as the location provided via the API?

If we always did this, it would help clear up the ambiguity for me of when we do and don't choose to accept information via the API. When we need to accept duplicated information like the package name for practical reasons (to support legacy packages), we still uphold the principle that the manifest is the source of truth.

**Comment by thomashoneyman** [comment]
> > 1. You commit and tag a new version in which you change the `repository` field in your manifest.
> > 2. You transfer the repository to its new location (for example, you move your repo from your account to `purescript-contrib`).
> > 3. You ask the registry to perform an Update, giving it the old repository. The registry will see that the old repository has a different repo specified in the manifest, so it will go fetch that repo and publish it at the ref you asked for.
> 
> After step (1) the Registry already has picked up the change in the URL when you publish this version (I assume you meant "publish" rather than "tag" - we don't need git tags at all! We pick the version from the `Manifest`), so after step (2) the new owners are ready to publish from the new location. There's no step (3) (also: we don't accept a `URL` in the `Update` operation)
>
> Since there's no step 3, we are looking at only one place and only one version published for doing the change. This is about the shortest protocol I can think of in any case? (short of already having keys available and going through another channel)

This wasn't a great example. Let me try to rephrase below, and please let me know if I'm getting the process wrong in any of these examples.

As a brief aside, when I said "commit and tag" I just meant "commit". You must commit your changes and push to your repository before asking the registry to publish, because the registry relies on the ref.

When transferring a repository, these actions must take place:

* Change the location in the manifest from the old to the new location, commit, and push the commit
* Transfer the repository from the old location to the new location
* Publish the commit via an `Update` operation

There are problems with this workflow depending on the order of these actions. (As a side note, we don't control this ordering whatsoever -- that's under the user's control.) Below, I'll walk through what happens in each scenario, as I understand it.

Note that in all cases, when the registry receives an `Update` it will contain the package name and ref to publish. The registry will look up the package name in the version metadata, and it will find the old location there. It will fetch the repository at the old location for the sake of publishing.

So what are the various scenarios?

**Order 1: Commit, publish, transfer**

The first way this can happen is that the user commits the changed manifest and publishes, and then transfers the repository. The registry looks up the old location, finds the updated manifest there, and sees that the URL has changed. But then it gets ambiguous:

* Should the registry publish the package from the old location? This will succeed, but it will result in a manifest file that states it was packaged from the new location, even though it was actually packaged from the old location. If the user never ends up transferring their repository, or they had a typo, then this is especially broken! The registry will only look at the new location if they try to publish anything else.
* Should the registry publish the package from the new location? This cannot succeed, because the repository hasn't been transferred yet. The only way this would succeed is the way I originally described: the user has to have the repository available in both the old and new locations at the same time, which we can't really ask for.

**Order 2: commit, transfer, publish || transfer, commit, publish**

The second way this can happen is that the repository transfer happens before the registry publishes the updated manifest. For example, the user commits a new location to the manifest, transfers the repo, and publishes, or the user transfers the repo, commits the new location, and publishes. Either way, this causes issues for the registry:

* The registry can't package the repository from the old location, because the old location doesn't necessarily exist anymore -- the package has been transferred.
* The registry can't package the repository from the new location, because the registry doesn't know what the new location is -- it's been committed to the manifest, but the manifest has already been transferred to the new location.

In the case of GitHub, perhaps this isn't an issue, because GitHub automatically sets up redirects. But since we are almost completely agnostic to the source of the tarball, we can't rely on this behavior.

---

This is why I think we need the user to authenticate a repository transfer. (Unless I've misrepresented the process!) Let me know if this makes sense or not!

**Comment by thomashoneyman** [comment]
> I can't quite explain why, but it feels wrong to not store in the data itself the location where the data was fetched from

I'm sympathetic to this. I am happy to keep it in there, so long as it's guaranteed to be correct (ie. we promise to verify it). For example, the rule I proposed [in this comment](https://github.com/purescript/registry/issues/331#issuecomment-1033045385):

> > What do you think of following a rule where if we receive information via the API that is also present in the manifest, we must verify that the two fields match?

**Comment by thomashoneyman** [comment]
> This is why I think we need the user to authenticate a repository transfer. (Unless I've misrepresented the process!) Let me know if this makes sense or not!

As discussed in the registry call this morning, this of course only works if they set their public key in the `owners` field. If they forgot that, then transfer the repo, they're still out of luck.

**Edit**: So, to summarize, my current understanding of the situation is this:

1. If a user transfers their repository without first setting their public key in the owners field or publishing a version that updates the manifest to the new location, then they will not be able to use the registry to publish further versions. They will have to transfer back to the old location and perform one of these steps first. If they can't do that, then it's unspecified what we will do to remedy the situation.
2. If a user publishes a new version in which they update their manifest to point at the new location they are going to transfer to, then the registry will create a tarball for their package from its current location, and will update its metadata to look at the new location for all future releases. If the package source is never transferred to the new location, then they will not be able to use the registry to publish further versions. It's unspecified what we will do to remedy the situation.
3. If a user publishes their public key in the owners field of their manifest prior to transferring to the new location, then they can use their SSH key to sign a transfer operation to the registry asserting that the location has changed. If the package source is never transferred, they can sign another request transferring it back to the old location.

I think we still need to decide what to do in these "unspecified" situations I've described above. Do you have any ideas, @f-f?

**Comment by thomashoneyman** [comment]
> I have been designing while trying to keep agnostic to the protocol we use to fetch stuff, even though in many places we do say "git" rather than "whatever protocol" (names are hard), so we could do a pass removing this terminology if you'd like

@f-f I have opened #345 to discuss this specifically, so we can stick to just the repository transfer topic in this thread to stay organized, if that works for you!

**Comment by f-f** [comment]
> I'm trying to clarify what exactly we do require of packages. Would you say that these are accurate requirements we impose on your package?

I think it's a bit more liberal than that: we can basically go fetch anything out of a key-value store from a `location`, over any protocol as long as it's decent to implement support for it, where the `ref` is the key for looking up the right sources.

> I'm sympathetic to this. I am happy to keep it in there, so long as it's guaranteed to be correct (ie. we promise to verify it). For example, the rule I proposed https://github.com/purescript/registry/issues/331#issuecomment-1033045385:

My understanding is that you proposed a different thing there: matching the _API_ input with the `Manifest` content.
I don't think that is very useful (since we don't trust the API input anyways), except for the case of the package name for non-legacy packages, which might make sense.

I now think though that we should always match the content of the `Metadata` and the info we get from each new `Manifest` (e.g. keeping the `repository`, `owners` fields in sync (I guess we'd match the repository, but update the owners?).
If we were to do that, then "changing the repo by changing the URL" would not be feasible anymore, but we'd have to do it through the SSH auth.
I'm fine with this - after all it's easier to just have one path for authenticated things to go through.

> I think we still need to decide what to do in these "unspecified" situations I've described above. Do you have any ideas?

Yes :smile: 
1. out of luck
2. out of luck

**Comment by thomashoneyman** [comment]
> If a user publishes a new version in which they update their manifest to point at the new location they are going to transfer to, then the registry will create a tarball for their package from its current location, and will update its metadata to look at the new location for all future releases. If the package source is never transferred to the new location, then they will not be able to use the registry to publish further versions. It's unspecified what we will do to remedy the situation.

This is the worst possible scenario, to my mind -- you actually get locked out of the registry altogether, via a simple mistake! But if we do this:

> If we were to do that, then "changing the repo by changing the URL" would not be feasible anymore, but we'd have to do it through the SSH auth. I'm fine with this - after all it's easier to just have one path for authenticated things to go through.

...then this is no longer possible. The worst possible case then is that you transfer the repository without setting an `owners` key, but then can't publish from the new location, and can't transfer back to the old location, at which point your only option is to try and work things out with the registry trustees and you may lose the ability to publish the package to the registry. And that, to me, is a workable edge case.

So a 👍 from me on unifying to a single SSH authentication operation for repo transfers.

**Comment by thomashoneyman** [comment]
> I think it's a bit more liberal than that: we can basically go fetch anything out of a key-value store from a location, over any protocol as long as it's decent to implement support for it, where the ref is the key for looking up the right sources.

Cool -- I'll rely on this phrasing. Thanks!

**Comment by thomashoneyman** [comment]
> I now think though that we should always match the content of the Metadata and the info we get from each new Manifest (e.g. keeping the repository, owners fields in sync (I guess we'd match the repository, but update the owners?).

Yea, it's a little funky because some values should only be changeable via an authenticated operation (ie. the `location` field), while others should be changeable by changing the value in the manifest directly (ie. the `owners` field). So we trust the manifest and the metadata for different things.

We'll need to decide what fields should be prioritized in each file (we trust the metadata for the location and the manifest for the owners, for example), and then decide how to resolve conflicts (do we always take the trusted value, or do we error if a package has a mismatch so they can fix the mismatch?).

Then there's the _separate_ issue of when to trust the API vs. trust the manifest. My understanding is that in an `Addition` we take the package name from the API call, because of the legacy use case, and otherwise our policy is "don't trust the API." This is minor, but maybe we should have an explicit 'legacy' addition after all, so that non-legacy packages don't send the package name along and we just take it from the manifest? More effort than it's worth?

I'm just trying to minimize the number of things we have to cross-reference, and eliminate multiple sources of information if possible so that we don't inadvertently make the wrong decision. It's hard to keep all of this in mind while developing!

**Comment by f-f** [comment]
> at which point your only option is to try and work things out with the registry trustees and you may lose the ability to publish the package to the registry

Being flexible about this means opening the door to social engineering. I think the policy on this should be inflexible: if you can't sign anything with the keys we have, then we don't know who you are and we can't do anything about it.

> So we trust the manifest and the metadata for different things

It's a very specific set of things: at this point we trust the `Manifest` for updating the `owners` field in the `Metadata`, and trust the metadata for everything else.

> This is minor, but maybe we should have an explicit 'legacy' addition after all, so that non-legacy packages don't send the package name along and we just take it from the manifest? More effort than it's worth?

I think this is alright, we have a reliable way to detect legacy packages without having to differentiate in the API, so yeah, changing this again feels like more effort than it's worth.

**Comment by thomashoneyman** [comment]
> then decide how to resolve conflicts (do we always take the trusted value, or do we error if a package has a mismatch so they can fix the mismatch?).

How do you feel about this?

**Comment by f-f** [comment]
> How do you feel about this?

`owners` is always updated, but a mismatch of package name (except for legacy packages), location and possibly other stuff should error out.
The package manager should run all these checks locally anyways before allowing a publish.

**Comment by thomashoneyman** [comment]
This discussion has now been specced as part of the `Transfer` operations section.
